plugins/statistics.koplugin/main.lua

```diff
@@ -73,58 +83,350 @@ function ReaderStatistics:init()
+                    book_stats.highlights = old_data.highlights
+                    book_stats.pages = old_data.pages
+                    book_stats.series = old_data.series
+                    book_stats.language = old_data.language
+                    book_stats.file = nil
+                    if self:addBookStatToDB(book_stats, conn) then
+                        nr_of_conv_books = nr_of_conv_books + 1
+                    else
+                        logger.dbg("Book not converted (old stats): " .. book_stats.title)
+                    end
+                end
+            end
+        end
+    end
+    local stmt = conn:prepare("INSERT INTO info values (?)")
+    stmt:reset():bind("1"):step()
+    stmt:close()
+    conn:close()
+    return nr_of_conv_books
+end
+
+function ReaderStatistics:getIdBookDB()
+    local conn = SQ3.open(db_location)
+    local id_book
+    local book_title = self.data.title:gsub("'", "''")
+    local book_authors = self.data.authors:gsub("'", "''")
+    local res = conn:rowexec(string.format("SELECT COUNT(id) FROM book WHERE title = '%s' AND authors = '%s' AND md5 = '%s';)",
+        book_title, book_authors, self.data.md5))
+    if res == 0 then
+        local stmt = conn:prepare("INSERT INTO book VALUES(NULL, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)")
+        stmt:reset():bind(self.data.title, self.data.authors, self.data.notes,
+            TimeVal:now().sec, self.data.highlights, self.data.pages,
+            self.data.series, self.data.language, self.data.md5, self.curr_total_time, self.curr_total_pages):step()
+        id_book = conn:rowexec("SELECT last_insert_rowid() AS num;")
+    else
+        id_book = conn:rowexec(string.format("SELECT id FROM book WHERE title = '%s' AND authors = '%s' AND md5 = '%s';)",
+            book_title, book_authors, self.data.md5))
+    end
+    conn:close()
+    return tonumber(id_book)
+end
+
+function ReaderStatistics:insertDB(id_book)
+    self.pages_stats[TimeVal:now().sec] = self.curr_page
+    if id_book == nil or util.tableSize(self.pages_stats) < 2 then
+        return
+    end
+    local diff_time
+    local conn = SQ3.open(db_location)
+    local sorted_performance = {}
+    for k, v in pairs(self.pages_stats) do
+        table.insert(sorted_performance, k)
+    end
+    table.sort(sorted_performance)
+    conn:exec('BEGIN')
+    local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat VALUES(?, ?, ?, ?)")
+    for i=1, #sorted_performance - 1 do
+        diff_time = sorted_performance[i+1] - sorted_performance[i]
+        if diff_time <= self.page_max_read_sec and diff_time >= self.page_min_read_sec  then
+            stmt:reset():bind(id_book, self.pages_stats[sorted_performance[i]], sorted_performance[i], diff_time):step()
+        elseif diff_time > self.page_max_read_sec then
+            stmt:reset():bind(id_book, self.pages_stats[sorted_performance[i]], sorted_performance[i], self.page_max_read_sec):step()
```

Comment:
Why do you save a value (page_max_read_sec) in case when a user did not read a page for page_max_read_sec minutes? If a user did not read a page, then, he will reread the same page or skip it. In this case, when user reread the same page you will add incorrect value page_max_read_sec and an actual correct read value.